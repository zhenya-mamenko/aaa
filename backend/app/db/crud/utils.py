import json
import re

from pydantic import BaseModel as _BaseModel

from app.db.sqlite import (
    execute_query as _execute_query,
    get_database_conn as _get_database_conn,
    read_query as _read_query,
)


def create_parameters_from_model(
    model: _BaseModel,
    /,
    exclude_fields: tuple[str] | None = None,
    only_fields: tuple[str] | None = None,
    get_max_data: tuple | None = None,
) -> tuple:
    """
    Create SQL parameters from a Pydantic model.

    Args:
        model (_BaseModel): The Pydantic model instance.
        exclude_fields (tuple[str] | None, optional): Fields to exclude from the model. Defaults to None.
        only_fields (tuple[str] | None, optional): Fields to include from the model. Defaults to None.
        get_max_data (tuple | None, optional): Tuple containing table name and field to get max value. Defaults to None.

    Returns:
        tuple: A tuple containing the SQL template, fields, parameters, and optionally a max query.
    """

    exclude_fields = exclude_fields or ()
    model_dump = {
        k: json.dumps(v) if type(v).__name__ in ["list", "dict"] else v for k, v in model.model_dump().items()
    }

    if only_fields:
        fields = [field for field in only_fields if field in model_dump]
        if len(fields) != len(only_fields):
            raise ValueError(f"Fields={set(fields) ^ set(only_fields)} are not in the model.")
    else:
        fields = [field for field, value in model_dump.items() if field not in exclude_fields and value is not None]
    parameters = tuple([model_dump[field] for field in fields])
    template = ", ".join(["?" for _ in parameters])

    if get_max_data:
        table_name, field = get_max_data
        if field in fields:
            raise ValueError(f"Field {field} must be excluded from the model.")
        fields.append(field)
        template += f", (SELECT COALESCE(MAX({field}), 0) + 1 FROM {table_name})"
        max_query = f"SELECT MAX({field}) as {field} FROM {table_name};"
        return template, fields, parameters, max_query

    return template, fields, parameters


def insert_data_from_model(
    *, table_name: str, id_field: tuple | str, model: _BaseModel, auto_increment_id: bool = True
) -> int | str | None:
    """
    Insert data into a table from a Pydantic model.

    Args:
        table_name (str): The name of the table to insert data into.
        id_field (tuple | str): The field(s) representing the ID.
        model (_BaseModel): The Pydantic model instance.
        auto_increment_id (bool, optional): Whether the ID is auto-incremented. Defaults to True.

    Returns:
        int | str | None: The ID of the inserted row if auto_increment_id is True, otherwise None.
    """

    if auto_increment_id:
        if isinstance(id_field, tuple):
            raise ValueError("auto_increment_id must be False if id_field is a tuple.")
        template, fields, parameters, q = create_parameters_from_model(model, get_max_data=(table_name, id_field))
    else:
        template, fields, parameters = create_parameters_from_model(model)

    with _get_database_conn() as conn:
        query = f"""
            INSERT INTO {table_name} (
                {", ".join(fields)}
            ) VALUES (
                {template}
            );
        """
        _execute_query(query, parameters=parameters, conn=conn)
        if auto_increment_id:
            id = _read_query(q, conn=conn)[0][id_field]
            return id

    return None


def update_data_from_model(*, table_name: str, id_field: tuple | str, model: _BaseModel) -> bool:
    """
    Update data in a table from a Pydantic model.

    Args:
        table_name (str): The name of the table to update data in.
        id_field (tuple | str): The field(s) representing the ID.
        model (_BaseModel): The Pydantic model instance.

    Returns:
        bool: True if the update was successful, False otherwise.
    """

    if isinstance(id_field, str):
        id_field = (id_field,)

    _, fields, parameters = create_parameters_from_model(model, exclude_fields=id_field)

    query = f"""
        UPDATE {table_name}
        SET
            {", ".join([f"{field} = ?" for field in fields])}
        WHERE
            {' AND '.join([f'{f} = ?' for f in id_field])};
    """
    parameters += tuple([getattr(model, f) for f in id_field])
    result = _execute_query(query, parameters=parameters)
    return result == 1


def delete_data_by_id(*, table_name: str, id_field: tuple | str, id_value: int | str | tuple) -> bool:
    """
    Delete data from a table by ID.

    Args:
        table_name (str): The name of the table to delete data from.
        id_field (tuple | str): The field(s) representing the ID.
        id_value (int | str | tuple): The value(s) of the ID to delete.

    Returns:
        bool: True if the deletion was successful, False otherwise.
    """

    if not isinstance(id_field, tuple):
        id_field=(id_field, )
    query = f"""
        DELETE FROM {table_name}
        WHERE
            {' AND '.join([f'{f} = ?' for f in id_field])};
    """
    if not isinstance(id_value, tuple):
        id_value=(id_value, )
    result = _execute_query(query, parameters=id_value)

    return result == 1


def extract_create_stmts_objects(sql_file_path: str, object_type: str) -> list[str]:
    """
    Extract the CREATE statements objects from a SQL file.

    Args:
        sql_file_path (str): The path to the SQL file.
        object_type (str): The type of objects to extract from the SQL file.

    Returns:
        list: A list of dictionaries containing the objects.
    """

    try:
        with open(sql_file_path, "r") as f:
            sql = f.read()
    except FileNotFoundError:
        raise FileNotFoundError(f"File not found: {sql_file_path}")

    re_template = fr"CREATE\s+{object_type}\s+(IF\s+NOT\s+EXISTS\s+)*(\S+)"
    pattern = re.compile(re_template)
    result = []
    for m in pattern.finditer(sql):
        result.append(m.group(2))

    return result
