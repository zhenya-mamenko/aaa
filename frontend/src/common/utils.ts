import { type AppElement } from "@/types/common";
import { elements } from "@/common/const";


export const isPlainObject: Function = (o: any) => Boolean(
  o && o.constructor && o.constructor.prototype && o.constructor.prototype.hasOwnProperty("isPrototypeOf")
);


export const flattenObject: Function = (obj: any, keys = []) => {
  return Object.keys(obj).reduce((acc: any, key: any) => {
    return Object.assign(acc, isPlainObject(obj[key])
      ? flattenObject(obj[key], keys.concat(key))
      : { [keys.concat(key).join(".")]: obj[key] }
    )
  }, {})
}


export const loadParams: Function = (key: string) => {
  const savedParams = localStorage.getItem(key)
  if (savedParams) {
    return JSON.parse(savedParams)
  }
  return null
}


export const saveParams: Function = (key: string, params: any) => {
  localStorage.setItem(key, JSON.stringify(params))
}


export const deleteParams: Function = (key: string) => {
  localStorage.removeItem(key)
}


export const removeReactivity: Function = (o: any) => {
  return JSON.parse(JSON.stringify(o))
}

export const findElement: Function = (key: string): AppElement | undefined => {
  return elements.find((element: AppElement) => element.key === key);
}


// String.formatUnicorn polyfill from https://stackoverflow.com/a/18234317
export const format: Function = function(str: string, ...rest: any[]): string {
    if (rest.length) {
      var t = typeof rest[0];
      var key;
      var args = ("string" === t || "number" === t) ? Array.prototype.slice.call(rest) : rest[0];
      for (key in args) {
        str = str.replace(new RegExp("\\{" + key + "\\}", "gi"), args[key]);
      }
    }
    return str;
  };


export const parseAxiosError: Function = (error: any): string | any[] => {
  if (error.response?.data?.detail !== undefined) {
    let message = "";
    const detail = error.response.data.detail;
    if (typeof detail === "string" && detail.length > 0) {
      message = `messages.backend.${detail.split(":")[0].replace(/ /g, "_")}`;
    }
    return [
      "messages.backend.request_error",
      error.response.status,
      message,
    ];
  } else if (!error.response) {
    return "messages.backend.no_reponse";
  } else {
    return error.message || "messages.backend.unknown_error";
  }
}


export const calculateReplenishment: Function = (
  list: any[],
  summa: number,
  percentileKey: string = "structure_percentile",
  resultKey: string = "calculated"
): void => {
  list.forEach(item => {
    (item as any)[resultKey] = Math.round(summa * item[percentileKey] / 1000);
  });
}


export const calculateRebalancing: Function = (
  list: any[],
  summa: number,
  amountKey: string = "amount",
  percentileKey: string = "structure_percentile",
  resultKey: string = "calculated"
): void => {
  const total = list.reduce((acc, item) => acc + item[amountKey], 0) + summa * 100;
  list.forEach(item => {
    (item as any)[resultKey] = Math.round(((total * item[percentileKey] / 1000) - item[amountKey]) / 100);
  });
}
