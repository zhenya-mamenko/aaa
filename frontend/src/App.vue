<template>
  <v-app full-height :theme="theme || 'light'">
    <v-app-bar color="primary" scroll-behavior="fully-hide">
      <v-app-bar-nav-icon data-testid="app-icon"
        variant="text"
        @click.stop="drawer = !drawer"
      />
      <v-app-bar-title data-testid="app-title">
        {{ t("app_title") }}
      </v-app-bar-title>
      <v-btn data-testid="app-configure"
        icon="mdi-cog"
        @click="showSettings"
      />
    </v-app-bar>
    <v-navigation-drawer
      color="primary"
      data-testid="app-navigation-drawer"
      :temporary="true"
      v-model="drawer"
    >
      <v-list data-testid="app-menu"
        :items="menu"
        @click:select="menuSelected"
      />
    </v-navigation-drawer>
    <v-main>
      <v-container class="h-100">
        <v-row class="h-100">
          <v-col class="h-100 max" cols="12" lg="10" offset-lg="1" xl="8" offset-xl="2" xxl="6" offset-xxl="3">
            <component :is="pageComponent" v-bind="pageProps"/>
          </v-col>
        </v-row>
        <form-dialog
          :opened="dialog.opened" :params="dialog.params" :title="dialog.title"
          @close="closeDialog" @reset="resetSettings" @save="saveSettings"
        />
      </v-container>
    </v-main>
  </v-app>
</template>

<script lang="ts">
import { nextTick } from "vue";
import { state } from "@/common/state";
import { t, setLocale } from "@/common/locale";
import { deleteParams, findElement, loadParams, removeReactivity, saveParams } from "@/common/utils";
import { defaultSettings } from "@/common/const";

import { type Settings, type AppElement } from "@/types/common";

import ds from "@/services/ConfigDataService";
import { type ConfigEntry } from "@/types/Config";

import AssetsState from "@/components/AssetsState.vue";
import AssetsValues from "@/components/AssetsValues.vue";
import DictionaryPage from "@/components/DictionaryPage.vue";
import PortfolioPage from "@/components/PortfolioPage.vue";
import StructureCategories from "./components/StructureCategories.vue";


export default {
  computed: {
    menu() {
      const menuItem = (element: AppElement) => ({
        title: t(element.title),
        value: element.key,
      });

      return [
        menuItem(findElement("portfolio")),
        menuItem(findElement("assets_state")),
        menuItem(findElement("statements")),
        { type: "divider" },
        {
          title: t("dictionaries"),
          children: [
            menuItem(findElement("structure_categories")),
            { type: "divider" },
            menuItem(findElement("assets")),
            menuItem(findElement("categories")),
            menuItem(findElement("classes")),
            menuItem(findElement("structures")),
          ]
        },
        { type: "divider" },
        menuItem(findElement("settings")),
      ]
    },
    pageComponent() {
      const components: { [key: string]: any } = {
        assets_state: AssetsState,
        assets: DictionaryPage,
        categories: DictionaryPage,
        classes: DictionaryPage,
        statements: AssetsValues,
        structures: DictionaryPage,
        structure_categories: StructureCategories,
        portfolio: PortfolioPage,
      }
      return components[this.page] ?? components["portfolio"];
    },
    t() {
      return t;
    },
    theme() {
      return state.theme;
    },
  },
  data() {
    return {
      dialog: {
        isGlobal: false,
        opened: false,
        params: {} as Settings,
        title: "",
      },
      drawer: false,
      page: "portfolio",
      pageProps: {} as { [key: string]: any },
    }
  },
  methods: {
    closeDialog() {
      this.dialog.opened = false;
      this.dialog.isGlobal = false;
    },
    menuSelected(event: any) {
      this.drawer = false;
      const item = findElement(event.id);
      if (item.type === "page" || item.type === "complex_dictionary") {
        this.pageProps = {};
        this.page = item.key;
      } else if (item.key === "settings") {
        this.dialog.isGlobal = true;
        this.showSettings();
      } else if (item.type === "dictionary") {
        this.page = item.key;
        this.pageProps.pageKey = item.key;
      }
    },
    resetSettings() {
      this.dialog.opened = false;
      const key = this.dialog.isGlobal ? "global" : this.page;
      deleteParams(key);
      state.settings[key] = removeReactivity(defaultSettings[key]);
      if (this.dialog.isGlobal) {
        this.dialog.isGlobal = false;
        this.updateGlobalSettings();
      }
    },
    saveSettings(newValue: any) {
      this.dialog.opened = false;
      const key = this.dialog.isGlobal ? "global" : this.page;
      saveParams(key, newValue);
      state.settings[key] = newValue;
      if (this.dialog.isGlobal) {
        this.dialog.isGlobal = false;
        this.updateGlobalSettings();
      }
    },
    showSettings() {
      const key = this.dialog.isGlobal ? "global" : this.page;
      this.dialog.params = removeReactivity(state.settings[key]);
      if (this.dialog.isGlobal) {
        this.dialog.title = t("settings.title");
      } else {
        this.dialog.title = `${t("settings.title")}<span class="d-none d-md-inline"> : ${t(findElement(this.page).title)}</span>`;
      }
      this.dialog.opened = true;
    },
    async updateGlobalSettings() {
      state.settings["global"] = loadParams("global") || state.settings["global"];
      if (this.$vuetify.locale.current !== state.locale) {
        this.$vuetify.locale.current = state.locale;
      }
      setLocale(state.locale);
      const data: ConfigEntry = {
        config_name: "currency",
        config_value: {
          symbol: state.currencySymbol,
          position: state.currencyPosition,
        },
      }
      if (!!ds && !!data) {
        await ds.updateOrCreate(data);
      }
      await nextTick();
      if (state.refreshCallbacks[this.page]) {
        await state.refreshCallbacks[this.page]();
      }
    },
  },
  mounted() {
    this.updateGlobalSettings();
    setLocale(state.locale);
    document.title = t("app_title");
  },
}
</script>

<style>
  .max {
    max-height: calc(100vh - 75px) !important;
  }
  .v-data-table__th {
    border-top: 1px solid #546E783F !important;
    border-bottom: 2px solid #546E782F !important;
  }
</style>
