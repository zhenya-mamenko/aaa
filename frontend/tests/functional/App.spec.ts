import "../localStorageMock";
import { mountComponent, appStubs, baseStubs, } from "./utils";
// @ts-ignore
import App from "@/App.vue";


jest.mock("@/common/locale", () => ({
  t: (key: string) => key,
  setLocale: (locale: string) => {},
}));

jest.mock("@/services/ConfigDataService", () => ({
  updateOrCreate: () => Promise.resolve({ data: {}, status: 200, }),
}));

jest.mock("@/services/DataService", () => ({
  getPortfolio: () => Promise.resolve({ data: [], }),
}));


describe("App", () => {

  function mount(data: any) {
    return mountComponent(
      App,
      {
        simpleStubs: ["v-btn", "v-divider", "v-spacer", "form-dialog", "v-data-table", "v-toolbar", ],
        stubs: [...baseStubs, ...appStubs, "PortfolioPage", ],
      },
      {},
      {
        drawer: false,
        page: "portfolio",
        pageProps: {},
        ...data,
      },
    );
  }

  it("Should render", () => {
    const wrapper = mount({});

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should contain the menu", () => {
    const wrapper = mount({});

    const menuItems = wrapper.findAll("[data-testid='app-menu'] div.v-list-item");
    expect(menuItems.length).toBe(10);

    const itemTexts = menuItems.map((item) => item.text());
    [
      "portfolio.title", "assets.state", "assets.statements", "dictionaries", "structures.assets",
      "assets.title", "categories.title", "classes.title", "structures.title", "settings.title",
    ].forEach((text, index) => {
      expect(itemTexts[index]).toContain(text);
    });
  });

});
