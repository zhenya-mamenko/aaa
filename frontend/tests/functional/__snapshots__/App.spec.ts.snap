// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should render 1`] = `
"<div full-height="" theme="lightTheme">
  <div color="primary" scroll-behavior="fully-hide">
    <v-app-bar-nav-icon-stub data-testid="app-icon" variant="text"></v-app-bar-nav-icon-stub>
    <div data-testid="app-title">app_title</div>
    <v-btn-stub data-testid="app-configure" icon="mdi-cog"></v-btn-stub>
  </div>
  <div color="primary" data-testid="app-navigation-drawer" temporary="true" modelvalue="false">
    <div class="v-list v-theme--light v-list--density-default v-list--one-line" tabindex="0" role="listbox" data-testid="app-menu">
      <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
        <!---->
        <div class="v-list-item__content" data-no-activator="">
          <div class="v-list-item-title">portfolio.title</div>
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
        <!---->
        <div class="v-list-item__content" data-no-activator="">
          <div class="v-list-item-title">assets.state</div>
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
        <!---->
        <div class="v-list-item__content" data-no-activator="">
          <div class="v-list-item-title">assets.statements</div>
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <v-divider-stub inset="false" vertical="false"></v-divider-stub>
      <div class="v-list-group">
        <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text v-list-group__header" tabindex="-2" id="v-list-group--id-Symbol(3)"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
          <!---->
          <div class="v-list-item__content" data-no-activator="">
            <div class="v-list-item-title">dictionaries</div>
            <!---->
            <!---->
          </div>
          <div class="v-list-item__append">
            <div density="default" icon="$expand"></div>
            <!---->
            <div class="v-list-item__spacer"></div>
          </div>
        </div>
        <transition-stub name="" appear="false" persisted="false" css="false">
          <div class="v-list-group__items" role="group" aria-labelledby="v-list-group--id-Symbol(3)" style="display: none;">
            <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
              <!---->
              <div class="v-list-item__content" data-no-activator="">
                <div class="v-list-item-title">structures.assets</div>
                <!---->
                <!---->
              </div>
              <!---->
            </div>
            <v-divider-stub inset="false" vertical="false"></v-divider-stub>
            <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
              <!---->
              <div class="v-list-item__content" data-no-activator="">
                <div class="v-list-item-title">assets.title</div>
                <!---->
                <!---->
              </div>
              <!---->
            </div>
            <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
              <!---->
              <div class="v-list-item__content" data-no-activator="">
                <div class="v-list-item-title">categories.title</div>
                <!---->
                <!---->
              </div>
              <!---->
            </div>
            <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
              <!---->
              <div class="v-list-item__content" data-no-activator="">
                <div class="v-list-item-title">classes.title</div>
                <!---->
                <!---->
              </div>
              <!---->
            </div>
            <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
              <!---->
              <div class="v-list-item__content" data-no-activator="">
                <div class="v-list-item-title">structures.title</div>
                <!---->
                <!---->
              </div>
              <!---->
            </div>
          </div>
        </transition-stub>
      </div>
      <v-divider-stub inset="false" vertical="false"></v-divider-stub>
      <div class="v-list-item v-list-item--link v-theme--light v-list-item--density-default v-list-item--one-line rounded-0 v-list-item--variant-text" tabindex="-2"><span class="v-list-item__overlay"></span><span class="v-list-item__underlay"></span>
        <!---->
        <div class="v-list-item__content" data-no-activator="">
          <div class="v-list-item-title">settings.title</div>
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
  </div>
  <div>
    <div class="v-container v-locale--is-ltr h-100">
      <div class="v-row h-100">
        <div class="v-col-lg-10 v-col-xl-8 v-col-xxl-6 offset-lg-1 offset-xl-2 offset-xxl-3 v-col-12 h-100 max">
          <v-data-table-stub data-testid="portfolio-page-data-table" class="h-100" headers="[object Object],[object Object],[object Object],[object Object]" hide-default-footer="" items="" items-per-page="100" loading="true" loading-text="messages.loading" sort-asc-icon="mdi-triangle-small-up" sort-desc-icon="mdi-triangle-small-down" sticky="true"></v-data-table-stub>
        </div>
      </div>
      <form-dialog-stub opened="false" params="[object Object]" title=""></form-dialog-stub>
    </div>
  </div>
</div>"
`;
